In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline
import itertools

import env

In [2]:
url = f'mysql+pymysql://{env.username}:{env.password}@{env.host}/curriculum_logs'
df = pd.read_sql('SELECT * FROM logs JOIN cohorts on logs.user_id = cohorts.id', url)

In [3]:
df.head()

date      time                                 path  user_id  \
0  2018-01-26  09:55:03                                    /        1   
1  2018-01-26  09:56:02                              java-ii        1   
2  2018-01-26  09:56:05  java-ii/object-oriented-programming        1   
3  2018-01-26  09:56:06   slides/object_oriented_programming        1   
4  2018-01-26  09:56:24            javascript-i/conditionals        2   

   cohort_id            ip  id      name      slack  start_date    end_date  \
0        8.0  97.105.19.61   1    Arches    #arches  2014-02-04  2014-04-22   
1        8.0  97.105.19.61   1    Arches    #arches  2014-02-04  2014-04-22   
2        8.0  97.105.19.61   1    Arches    #arches  2014-02-04  2014-04-22   
3        8.0  97.105.19.61   1    Arches    #arches  2014-02-04  2014-04-22   
4       22.0  97.105.19.61   2  Badlands  #badlands  2014-06-04  2014-08-22   

            created_at           updated_at deleted_at  program_id  
0  2016-06-14 19:52:26  2016-06-14 19:52:26       None           1  
1  2016-06-14 19:52:26  2016-06-14 19:52:26       None           1  
2  2016-06-14 19:52:26  2016-06-14 19:52:26       None           1  
3  2016-06-14 19:52:26  2016-06-14 19:52:26       None           1  
4  2016-06-14 19:52:26  2016-06-14 19:52:26       None           1

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 73739 entries, 0 to 73738
Data columns (total 15 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   date        73739 non-null  object 
 1   time        73739 non-null  object 
 2   path        73739 non-null  object 
 3   user_id     73739 non-null  int64  
 4   cohort_id   72405 non-null  float64
 5   ip          73739 non-null  object 
 6   id          73739 non-null  int64  
 7   name        73739 non-null  object 
 8   slack       73739 non-null  object 
 9   start_date  73739 non-null  object 
 10  end_date    73739 non-null  object 
 11  created_at  73739 non-null  object 
 12  updated_at  73739 non-null  object 
 13  deleted_at  0 non-null      object 
 14  program_id  73739 non-null  int64  
dtypes: float64(1), int64(3), object(11)
memory usage: 8.4+ MB


In [5]:
#change the dtypes to the correct types

In [6]:
df.user_id = df.user_id.astype('object')

In [7]:
df.cohort_id = df.cohort_id.astype('object')

In [8]:
df.id = df.id.astype('object')

In [9]:
df.program_id = df.program_id.astype('object')

1. Which lesson appears to attract the most traffic consistently across cohorts (per program)?

In [10]:
df.path.value_counts(ascending = False)

/                                                                                5056
index.html                                                                       1520
javascript-i                                                                     1486
spring                                                                           1401
java-iii                                                                         1237
                                                                                 ... 
strudents                                                                           1
users/2/edit                                                                        1
11-nlp/word-2-vec.md                                                                1
5-stats                                                                             1
content/php_i/intro-to-operators/incrementing-and-decrementing-operators.html       1
Name: path, Length: 1442, dtype: int64

A: It looks like lesson jacascript-i is the most referenced lesson with 1486 requests

2. Is there a cohort that referred to a lesson significantly more than other cohorts seemed to gloss over?


In [11]:
# I want to concatenate the time and date column
df['date'] = df['date'] + ' ' + df['time']

In [12]:
df.head()

date      time                                 path user_id  \
0  2018-01-26 09:55:03  09:55:03                                    /       1   
1  2018-01-26 09:56:02  09:56:02                              java-ii       1   
2  2018-01-26 09:56:05  09:56:05  java-ii/object-oriented-programming       1   
3  2018-01-26 09:56:06  09:56:06   slides/object_oriented_programming       1   
4  2018-01-26 09:56:24  09:56:24            javascript-i/conditionals       2   

  cohort_id            ip id      name      slack  start_date    end_date  \
0       8.0  97.105.19.61  1    Arches    #arches  2014-02-04  2014-04-22   
1       8.0  97.105.19.61  1    Arches    #arches  2014-02-04  2014-04-22   
2       8.0  97.105.19.61  1    Arches    #arches  2014-02-04  2014-04-22   
3       8.0  97.105.19.61  1    Arches    #arches  2014-02-04  2014-04-22   
4      22.0  97.105.19.61  2  Badlands  #badlands  2014-06-04  2014-08-22   

            created_at           updated_at deleted_at program_id  
0  2016-06-14 19:52:26  2016-06-14 19:52:26       None          1  
1  2016-06-14 19:52:26  2016-06-14 19:52:26       None          1  
2  2016-06-14 19:52:26  2016-06-14 19:52:26       None          1  
3  2016-06-14 19:52:26  2016-06-14 19:52:26       None          1  
4  2016-06-14 19:52:26  2016-06-14 19:52:26       None          1

In [30]:
df = df.drop(columns='time')

In [31]:
# change the df to a datetime index
df.date = pd.to_datetime(df.date)
df = df.set_index('date')

In [15]:
df.head()

date      time                                 path user_id  \
0  2018-01-26 09:55:03  09:55:03                                    /       1   
1  2018-01-26 09:56:02  09:56:02                              java-ii       1   
2  2018-01-26 09:56:05  09:56:05  java-ii/object-oriented-programming       1   
3  2018-01-26 09:56:06  09:56:06   slides/object_oriented_programming       1   
4  2018-01-26 09:56:24  09:56:24            javascript-i/conditionals       2   

  cohort_id            ip id      name      slack  start_date    end_date  \
0       8.0  97.105.19.61  1    Arches    #arches  2014-02-04  2014-04-22   
1       8.0  97.105.19.61  1    Arches    #arches  2014-02-04  2014-04-22   
2       8.0  97.105.19.61  1    Arches    #arches  2014-02-04  2014-04-22   
3       8.0  97.105.19.61  1    Arches    #arches  2014-02-04  2014-04-22   
4      22.0  97.105.19.61  2  Badlands  #badlands  2014-06-04  2014-08-22   

            created_at           updated_at deleted_at program_id  
0  2016-06-14 19:52:26  2016-06-14 19:52:26       None          1  
1  2016-06-14 19:52:26  2016-06-14 19:52:26       None          1  
2  2016-06-14 19:52:26  2016-06-14 19:52:26       None          1  
3  2016-06-14 19:52:26  2016-06-14 19:52:26       None          1  
4  2016-06-14 19:52:26  2016-06-14 19:52:26       None          1

In [16]:
# detect anomalies start by gettting the count
df.path.value_counts()

/                                                                                5056
index.html                                                                       1520
javascript-i                                                                     1486
spring                                                                           1401
java-iii                                                                         1237
                                                                                 ... 
strudents                                                                           1
users/2/edit                                                                        1
11-nlp/word-2-vec.md                                                                1
5-stats                                                                             1
content/php_i/intro-to-operators/incrementing-and-decrementing-operators.html       1
Name: path, Length: 1442, dtype: int64

In [17]:
# next find the frequency in the form of a percentage
df.path.value_counts(normalize=True).tail(50)

5-stats/2.5-distributions-and-testing                                                                                                                                                                                                                                      0.000014
prework/fundamentals                                                                                                                                                                                                                                                       0.000014
students/468/notes                                                                                                                                                                                                                                                         0.000014
students/643/This%20alumnus%20works%20for%20Appdiction%20Studios,%20as%20of%20this%20timestamp.%20This%20company%20is%20launching%20a%20bootcamp%20program%20in%20competitio

In [18]:
def value_counts_and_frequencies(s: pd.Series, dropna=True) -> pd.DataFrame:
    return pd.merge(
        s.value_counts(dropna=False).rename('count'),
        s.value_counts(dropna=False, normalize=True).rename('proba'),
        left_index=True,
        right_index=True,
    )

In [19]:
# create a dataframe
path_df = value_counts_and_frequencies(df.path)
path_df.head(25)

count     proba
/                                  5056  0.068566
index.html                         1520  0.020613
javascript-i                       1486  0.020152
spring                             1401  0.018999
java-iii                           1237  0.016775
java-ii                            1126  0.015270
appendix                           1000  0.013561
javascript-ii                       937  0.012707
mysql                               930  0.012612
java-i                              930  0.012612
html-css                            894  0.012124
jquery                              888  0.012042
mkdocs/search_index.json            789  0.010700
classification/overview             585  0.007933
search/search_index.json            577  0.007825
spring/fundamentals/repositories    577  0.007825
sql/mysql-overview                  555  0.007527
javascript-i/functions              550  0.007459
spring/fundamentals/views           529  0.007174
javascript-i/loops                  513  0.006957
mysql/tables                        511  0.006930
javascript-i/conditionals           508  0.006889
6-regression/1-overview             493  0.006686
mysql/databases                     491  0.006659
mysql/users                         476  0.006455

In [20]:
# check for missing values
df.isna().sum()

date              0
time              0
path              0
user_id           0
cohort_id      1334
ip                0
id                0
name              0
slack             0
start_date        0
end_date          0
created_at        0
updated_at        0
deleted_at    73739
program_id        0
dtype: int64

In [21]:
#remove missing values
df = df[~df['cohort_id'].isnull()]

In [22]:
df.isna().sum()

date              0
time              0
path              0
user_id           0
cohort_id         0
ip                0
id                0
name              0
slack             0
start_date        0
end_date          0
created_at        0
updated_at        0
deleted_at    72405
program_id        0
dtype: int64

In [23]:
#dropping the column deleted_at since I won't need that for anything
df = df.drop(columns='deleted_at')

In [24]:
df.shape

(72405, 14)

In [25]:
# conditional probabilities --- cohort given path
cohort_id_given_path = (
    df.groupby('path')
    .cohort_id.value_counts(normalize=True)
    .rename('proba_cohort_id_given_path')
    .reset_index()
)
cohort_id_given_path.head(30)

path  cohort_id  \
0                                          /       28.0   
1                                          /       22.0   
2                                          /       23.0   
3                                          /       21.0   
4                                          /        1.0   
5                                          /        8.0   
6                                          /       13.0   
7                                          /       14.0   
8                                          /       18.0   
9                                        00_       28.0   
10                                  00_index       28.0   
11                                  01_intro       28.0   
12                          02_listing_files       28.0   
13                             03_file_paths       28.0   
14              04_navigating_the_filesystem       28.0   
15         05_creating_files_and_directories       28.0   
16                           06_moving_files       28.0   
17                            07_more_topics       28.0   
18               0_Classification_Algorithms        8.0   
19    1-fundamentals/1-fundamentals-overview       28.0   
20    1-fundamentals/1-fundamentals-overview       22.0   
21    1-fundamentals/1-intro-to-data-science       28.0   
22  1-fundamentals/1.1-intro-to-data-science       28.0   
23  1-fundamentals/1.1-intro-to-data-science       22.0   
24  1-fundamentals/1.1-intro-to-data-science       14.0   
25  1-fundamentals/1.2-data-science-pipeline       28.0   
26  1-fundamentals/1.2-data-science-pipeline       22.0   
27          1-fundamentals/1.3-pipeline-demo       28.0   
28          1-fundamentals/1.3-pipeline-demo       22.0   
29         1-fundamentals/2.1-excel-overview       28.0   

    proba_cohort_id_given_path  
0                     0.458342  
1                     0.281718  
2                     0.068931  
3                     0.066334  
4                     0.060340  
5                     0.040360  
6                     0.008991  
7                     0.008392  
8                     0.006593  
9                     1.000000  
10                    1.000000  
11                    1.000000  
12                    1.000000  
13                    1.000000  
14                    1.000000  
15                    1.000000  
16                    1.000000  
17                    1.000000  
18                    1.000000  
19                    0.966667  
20                    0.033333  
21                    1.000000  
22                    0.990164  
23                    0.006557  
24                    0.003279  
25                    0.902439  
26                    0.097561  
27                    0.916667  
28                    0.083333  
29                    1.000000

In [26]:
# there are 1436 paths 

In [27]:
# look at classification/knn and class/log reg as possible anomalies
# there are many more that are accessed by cohort 28 only

In [28]:
df[df.path == 'classification/knn']

date      time                path user_id cohort_id  \
64759  2020-09-08 16:16:00  16:16:00  classification/knn      11      28.0   
64760  2020-09-08 16:16:00  16:16:00  classification/knn      11      28.0   
64808  2020-09-10 09:32:46  09:32:46  classification/knn      11      28.0   
65037  2020-09-15 14:43:59  14:43:59  classification/knn      11      28.0   
65038  2020-09-15 14:44:20  14:44:20  classification/knn      11      28.0   
65154  2020-09-17 09:45:03  09:45:03  classification/knn      11      28.0   
65209  2020-09-17 15:14:01  15:14:01  classification/knn      11      28.0   
70786  2021-02-22 14:04:11  14:04:11  classification/knn      11      28.0   
70800  2021-02-22 14:11:00  14:11:00  classification/knn      11      28.0   
70973  2021-02-25 17:00:38  17:00:38  classification/knn      11      28.0   
71005  2021-02-26 09:58:38  09:58:38  classification/knn      11      28.0   
71112  2021-02-28 16:29:42  16:29:42  classification/knn      11      28.0   
71213  2021-03-02 10:07:44  10:07:44  classification/knn      11      28.0   
71256  2021-03-03 09:08:14  09:08:14  classification/knn      11      28.0   
71309  2021-03-04 09:32:38  09:32:38  classification/knn      11      28.0   
71311  2021-03-04 09:32:47  09:32:47  classification/knn      11      28.0   

                   ip  id name slack  start_date    end_date  \
64759  76.185.145.231  11  Ike  #ike  2016-01-20  2016-05-12   
64760  76.185.145.231  11  Ike  #ike  2016-01-20  2016-05-12   
64808  76.185.145.231  11  Ike  #ike  2016-01-20  2016-05-12   
65037  76.185.145.231  11  Ike  #ike  2016-01-20  2016-05-12   
65038  76.185.145.231  11  Ike  #ike  2016-01-20  2016-05-12   
65154  76.185.145.231  11  Ike  #ike  2016-01-20  2016-05-12   
65209  76.185.145.231  11  Ike  #ike  2016-01-20  2016-05-12   
70786  76.185.155.178  11  Ike  #ike  2016-01-20  2016-05-12   
70800  76.185.155.178  11  Ike  #ike  2016-01-20  2016-05-12   
70973  76.185.155.178  11  Ike  #ike  2016-01-20  2016-05-12   
71005  76.185.155.178  11  Ike  #ike  2016-01-20  2016-05-12   
71112  76.185.131.226  11  Ike  #ike  2016-01-20  2016-05-12   
71213  76.185.155.178  11  Ike  #ike  2016-01-20  2016-05-12   
71256  76.185.155.178  11  Ike  #ike  2016-01-20  2016-05-12   
71309  76.185.155.178  11  Ike  #ike  2016-01-20  2016-05-12   
71311  76.185.155.178  11  Ike  #ike  2016-01-20  2016-05-12   

                created_at           updated_at program_id  
64759  2016-06-14 19:52:26  2016-06-14 19:52:26          1  
64760  2016-06-14 19:52:26  2016-06-14 19:52:26          1  
64808  2016-06-14 19:52:26  2016-06-14 19:52:26          1  
65037  2016-06-14 19:52:26  2016-06-14 19:52:26          1  
65038  2016-06-14 19:52:26  2016-06-14 19:52:26          1  
65154  2016-06-14 19:52:26  2016-06-14 19:52:26          1  
65209  2016-06-14 19:52:26  2016-06-14 19:52:26          1  
70786  2016-06-14 19:52:26  2016-06-14 19:52:26          1  
70800  2016-06-14 19:52:26  2016-06-14 19:52:26          1  
70973  2016-06-14 19:52:26  2016-06-14 19:52:26          1  
71005  2016-06-14 19:52:26  2016-06-14 19:52:26          1  
71112  2016-06-14 19:52:26  2016-06-14 19:52:26          1  
71213  2016-06-14 19:52:26  2016-06-14 19:52:26          1  
71256  2016-06-14 19:52:26  2016-06-14 19:52:26          1  
71309  2016-06-14 19:52:26  2016-06-14 19:52:26          1  
71311  2016-06-14 19:52:26  2016-06-14 19:52:26          1

- possible anomalies:

do these lessons (class/knn and class/log-reg no longer exist is that why it was accessed solely by cohort 28?


In [32]:
# going back to exploring each feature individually

In [33]:
# detect anomalies start by gettting the count
df.ip.value_counts()

97.105.19.61      27448
97.105.19.58      15101
76.185.145.231     4754
67.10.174.56       3115
76.185.131.226     2618
                  ...  
104.184.220.38        1
99.203.154.93         1
70.123.229.111        1
172.56.15.51          1
99.203.155.25         1
Name: ip, Length: 369, dtype: int64

In [34]:
# next find the frequency in the form of a percentage
df.ip.value_counts(normalize=True)

97.105.19.61      0.379090
97.105.19.58      0.208563
76.185.145.231    0.065658
67.10.174.56      0.043022
76.185.131.226    0.036158
                    ...   
104.184.220.38    0.000014
99.203.154.93     0.000014
70.123.229.111    0.000014
172.56.15.51      0.000014
99.203.155.25     0.000014
Name: ip, Length: 369, dtype: float64

In [35]:
# this tells me with ip address 97.105.19.61 doing .379 of the requests it is probably
# the Code up ip address

In [36]:
def value_counts_and_frequencies(s: pd.Series, dropna=True) -> pd.DataFrame:
    return pd.merge(
        s.value_counts(dropna=False).rename('count'),
        s.value_counts(dropna=False, normalize=True).rename('proba'),
        left_index=True,
        right_index=True,
    )

In [37]:
# create a dataframe
ip_df = value_counts_and_frequencies(df.ip)
ip_df.head(20)

count     proba
97.105.19.61     27448  0.379090
97.105.19.58     15101  0.208563
76.185.145.231    4754  0.065658
67.10.174.56      3115  0.043022
76.185.131.226    2618  0.036158
104.190.242.242   1631  0.022526
70.121.215.53     1061  0.014654
66.69.40.223       989  0.013659
76.185.155.178     922  0.012734
65.43.115.179      856  0.011822
67.10.161.55       651  0.008991
192.171.117.210    557  0.007693
72.191.50.176      523  0.007223
108.65.244.91      516  0.007127
104.53.201.99      447  0.006174
70.121.199.82      442  0.006105
162.232.155.19     377  0.005207
96.8.150.23        372  0.005138
24.155.251.203     365  0.005041
72.191.29.130      362  0.005000

In [38]:
# conditional probabilities
cohort_id_given_ip = (
    df.groupby('ip')
    .cohort_id.value_counts(normalize=True)
    .rename('proba_cohort_id_given_ip')
    .reset_index()
)
cohort_id_given_ip.head()

ip  cohort_id  proba_cohort_id_given_ip
0    100.35.67.162       21.0                       1.0
1     104.1.46.153       23.0                       1.0
2  104.129.196.183       22.0                       1.0
3  104.156.210.198       22.0                       1.0
4   104.180.86.144       23.0                       1.0

In [39]:
# looking for anomalies

cohort_id_given_ip[cohort_id_given_ip.proba_cohort_id_given_ip < 1].head(30)

ip  cohort_id  proba_cohort_id_given_ip
32      107.77.64.27       21.0                  0.800000
33      107.77.64.27       22.0                  0.200000
54     12.189.101.90       22.0                  0.921569
55     12.189.101.90       21.0                  0.078431
56       12.247.3.42       21.0                  0.681818
57       12.247.3.42       22.0                  0.318182
58       12.247.3.50       22.0                  0.766234
59       12.247.3.50       21.0                  0.233766
60     12.27.152.137       16.0                  0.500000
61     12.27.152.137       22.0                  0.500000
89      148.66.37.91       22.0                  0.769231
90      148.66.37.91       21.0                  0.230769
115  170.248.173.245       16.0                  0.523810
116  170.248.173.245       22.0                  0.476190
117  170.248.173.246       22.0                  0.875000
118  170.248.173.246       16.0                  0.125000
119  170.248.173.247       16.0                  0.833333
120  170.248.173.247       22.0                  0.104167
121  170.248.173.247       21.0                  0.062500
145  173.173.127.101       22.0                  0.893939
146  173.173.127.101       21.0                  0.106061
174  192.171.117.210       22.0                  0.558348
175  192.171.117.210       28.0                  0.156194
176  192.171.117.210       14.0                  0.118492
177  192.171.117.210       23.0                  0.062837
178  192.171.117.210        1.0                  0.055655
179  192.171.117.210       21.0                  0.046679
180  192.171.117.210       13.0                  0.001795
208     24.243.10.53       22.0                  0.661538
209     24.243.10.53       18.0                  0.338462

In [40]:
# let's look more into cohort_id

In [41]:
# detect anomalies start by gettting the count
df.cohort_id.value_counts()

28.0    25097
22.0    23840
14.0     8401
23.0     5612
21.0     3413
1.0      3252
8.0      1617
13.0      592
18.0      474
16.0      107
Name: cohort_id, dtype: int64

In [42]:
# next find the frequency in the form of a percentage
df.cohort_id.value_counts(normalize=True)

28.0    0.346620
22.0    0.329259
14.0    0.116028
23.0    0.077508
21.0    0.047138
1.0     0.044914
8.0     0.022333
13.0    0.008176
18.0    0.006547
16.0    0.001478
Name: cohort_id, dtype: float64

In [43]:
def value_counts_and_frequencies(s: pd.Series, dropna=True) -> pd.DataFrame:
    return pd.merge(
        s.value_counts(dropna=False).rename('count'),
        s.value_counts(dropna=False, normalize=True).rename('proba'),
        left_index=True,
        right_index=True,
    )

In [44]:
# create a dataframe
cohort_id_df = value_counts_and_frequencies(df.cohort_id)
cohort_id_df

count     proba
28.0  25097  0.346620
22.0  23840  0.329259
14.0   8401  0.116028
23.0   5612  0.077508
21.0   3413  0.047138
1.0    3252  0.044914
8.0    1617  0.022333
13.0    592  0.008176
18.0    474  0.006547
16.0    107  0.001478

In [45]:
# conditional probabilities
cohort_id_given_start_date = (
    df.groupby('start_date')
    .cohort_id.value_counts(normalize=True)
    .rename('proba_cohort_id_start_date')
    .reset_index()
)
cohort_id_given_start_date.head()

start_date  cohort_id  proba_cohort_id_start_date
0  2014-02-04       28.0                    0.752927
1  2014-02-04        8.0                    0.210383
2  2014-02-04       21.0                    0.036690
3  2014-06-04       22.0                    1.000000
4  2014-09-04       22.0                    1.000000

there seems to be an anomaly either line 3 or 4 because there can't be 100%  chance that the start date for cohort 22 was both June and August of 2014

In [46]:
# let's look more at start_date individually

In [47]:
# detect anomalies start by gettting the count
df.start_date.value_counts(ascending = False)

2016-01-20    17913
2020-01-13    12329
2014-02-04     7686
2017-07-31     2263
2020-12-07     1954
2018-07-23     1809
2015-06-05     1803
2014-11-18     1701
2014-09-04     1562
2014-06-04     1541
2017-06-05     1533
2017-03-27     1532
2021-01-25     1459
2015-02-03     1408
2019-02-04     1397
2021-03-15     1264
2019-03-18     1004
2018-03-05      969
2015-11-03      922
2019-01-22      894
2016-03-08      889
2014-10-20      692
2021-03-03      684
2019-05-28      670
2016-07-18      596
2019-09-16      592
2015-03-30      585
2020-07-20      517
2019-08-19      509
2015-09-22      483
2018-11-05      482
2018-09-17      474
2016-09-26      296
2019-11-04      234
2021-04-12      204
2018-01-08      199
2016-10-26      199
2020-03-23      182
2017-02-06      175
2020-02-24      170
2020-11-02      157
2019-07-15      151
2017-09-27      107
2020-02-03       95
2016-05-23       94
2018-05-29       26
Name: start_date, dtype: int64

In [48]:
# next find the frequency in the form of a percentage
df.start_date.value_counts(normalize=True)

2016-01-20    0.247400
2020-01-13    0.170278
2014-02-04    0.106153
2017-07-31    0.031255
2020-12-07    0.026987
2018-07-23    0.024984
2015-06-05    0.024902
2014-11-18    0.023493
2014-09-04    0.021573
2014-06-04    0.021283
2017-06-05    0.021173
2017-03-27    0.021159
2021-01-25    0.020151
2015-02-03    0.019446
2019-02-04    0.019294
2021-03-15    0.017457
2019-03-18    0.013866
2018-03-05    0.013383
2015-11-03    0.012734
2019-01-22    0.012347
2016-03-08    0.012278
2014-10-20    0.009557
2021-03-03    0.009447
2019-05-28    0.009254
2016-07-18    0.008231
2019-09-16    0.008176
2015-03-30    0.008080
2020-07-20    0.007140
2019-08-19    0.007030
2015-09-22    0.006671
2018-11-05    0.006657
2018-09-17    0.006547
2016-09-26    0.004088
2019-11-04    0.003232
2021-04-12    0.002817
2018-01-08    0.002748
2016-10-26    0.002748
2020-03-23    0.002514
2017-02-06    0.002417
2020-02-24    0.002348
2020-11-02    0.002168
2019-07-15    0.002085
2017-09-27    0.001478
2020-02-03 

In [49]:
# create a dataframe
start_date_df = value_counts_and_frequencies(df.start_date)
start_date_df.head()

count     proba
2016-01-20  17913  0.247400
2020-01-13  12329  0.170278
2014-02-04   7686  0.106153
2017-07-31   2263  0.031255
2020-12-07   1954  0.026987

In [50]:
# conditional probabilities
start_date_given_ip = (
    df.groupby('ip')
    .start_date.value_counts(normalize=True)
    .rename('proba_start_date_given_ip')
    .reset_index()
)
start_date_given_ip.head(25)

ip  start_date  proba_start_date_given_ip
0     100.35.67.162  2019-11-04                        1.0
1      104.1.46.153  2021-03-03                        1.0
2   104.129.196.183  2016-03-08                        1.0
3   104.156.210.198  2019-05-28                        1.0
4    104.180.86.144  2020-12-07                        1.0
5    104.182.15.214  2017-06-05                        1.0
6    104.184.220.38  2014-06-04                        1.0
7   104.190.242.242  2017-07-31                        1.0
8     104.191.1.102  2017-03-27                        1.0
9    104.191.12.108  2019-09-16                        1.0
10    104.53.201.99  2015-02-03                        1.0
11   104.53.206.121  2017-03-27                        1.0
12    104.53.215.21  2019-07-15                        1.0
13  107.152.102.138  2017-07-31                        1.0
14  107.181.165.202  2017-07-31                        1.0
15    107.204.132.6  2018-01-08                        1.0
16    107.204.52.24  2021-01-25                        1.0
17   107.77.100.125  2015-11-03                        1.0
18   107.77.100.127  2018-07-23                        1.0
19    107.77.100.15  2020-03-23                        1.0
20    107.77.100.45  2016-07-18                        1.0
21    107.77.100.47  2019-01-22                        1.0
22    107.77.100.84  2016-07-18                        1.0
23    107.77.100.89  2020-02-24                        1.0
24    107.77.100.91  2019-01-22                        1.0

In [51]:
# looking for anomalies
start_date_given_ip[start_date_given_ip.proba_start_date_given_ip < 1]

ip  start_date  proba_start_date_given_ip
32    107.77.64.27  2015-11-03                   0.800000
33    107.77.64.27  2014-10-20                   0.200000
52   108.65.244.91  2021-03-15                   0.709302
53   108.65.244.91  2018-03-05                   0.290698
55   12.189.101.90  2014-09-04                   0.333333
..             ...         ...                        ...
440   97.105.19.61  2019-07-15                   0.002805
441   97.105.19.61  2020-02-24                   0.002550
442   97.105.19.61  2020-03-23                   0.001967
443   97.105.19.61  2018-05-29                   0.000947
444   97.105.19.61  2019-11-04                   0.000510

[152 rows x 3 columns]

for these ip addresses with more than one start date any ip location other than 
97.105.19.61 which I determined was the Code up location may be anomalies, so let's look at just those other addresses

In [52]:
df.ip[df.ip != '97.105.19.61'].head(30)

date
2018-01-26 10:59:02      71.137.146.71
2018-01-26 10:59:06      71.137.146.71
2018-01-26 12:20:36    209.205.123.242
2018-01-26 12:20:41    209.205.123.242
2018-01-26 13:02:29      12.189.101.90
2018-01-26 13:02:29      12.189.101.90
2018-01-26 13:02:31      12.189.101.90
2018-01-26 13:02:31      12.189.101.90
2018-01-26 16:00:29    209.205.123.242
2018-01-26 16:00:59    209.205.123.242
2018-01-26 16:02:16    209.205.123.242
2018-01-26 18:21:28        12.247.3.42
2018-01-26 18:21:37        12.247.3.42
2018-01-26 18:23:01        12.247.3.42
2018-01-26 20:17:32        12.247.3.42
2018-01-26 20:17:50        12.247.3.42
2018-01-26 20:18:00        12.247.3.42
2018-01-26 20:18:32        12.247.3.42
2018-01-26 20:18:46        12.247.3.42
2018-01-26 20:18:51        12.247.3.42
2018-01-26 20:19:05        12.247.3.42
2018-01-26 20:20:12        12.247.3.42
2018-01-26 20:21:30        12.247.3.42
2018-01-26 20:21:32        12.247.3.42
2018-01-26 20:21:40        12.247.3.42
2018-01-26 20:21:42 

In [53]:
start_date_given_ip[start_date_given_ip.ip != '97.105.19.61']

ip  start_date  proba_start_date_given_ip
0      100.35.67.162  2019-11-04                        1.0
1       104.1.46.153  2021-03-03                        1.0
2    104.129.196.183  2016-03-08                        1.0
3    104.156.210.198  2019-05-28                        1.0
4     104.180.86.144  2020-12-07                        1.0
..               ...         ...                        ...
490    99.203.27.185  2016-09-26                        1.0
491    99.203.27.231  2016-01-20                        1.0
492     99.203.27.36  2015-06-05                        1.0
493     99.203.27.87  2016-09-26                        1.0
494     99.47.142.88  2015-11-03                        1.0

[452 rows x 3 columns]

In [54]:
# if I want to answer which cohorts are accessing which paths the most

In [55]:
df['path_1'] = df.path.str.split('/').str[0]

df['path_2'] = df.path.str.split('/').str[1]

In [56]:
df.head(50)

path  \
date                                                                     
2018-01-26 09:55:03                                                  /   
2018-01-26 09:56:02                                            java-ii   
2018-01-26 09:56:05                java-ii/object-oriented-programming   
2018-01-26 09:56:06                 slides/object_oriented_programming   
2018-01-26 09:56:24                          javascript-i/conditionals   
2018-01-26 09:56:41                                 javascript-i/loops   
2018-01-26 09:56:46                          javascript-i/conditionals   
2018-01-26 09:56:48                             javascript-i/functions   
2018-01-26 09:56:59                                 javascript-i/loops   
2018-01-26 09:58:26                             javascript-i/functions   
2018-01-26 09:59:22                           mkdocs/search_index.json   
2018-01-26 10:00:02  javascript-i/introduction/working-with-data-ty...   
2018-01-26 10:00:37                                                  /   
2018-01-26 10:00:39                                       javascript-i   
2018-01-26 10:00:44  javascript-i/introduction/working-with-data-ty...   
2018-01-26 10:00:49                javascript-i/introduction/variables   
2018-01-26 10:00:57                javascript-i/introduction/operators   
2018-01-26 10:01:12  javascript-i/introduction/working-with-data-ty...   
2018-01-26 10:02:02                             javascript-i/functions   
2018-01-26 10:02:21                             javascript-i/functions   
2018-01-26 10:02:42                             javascript-i/functions   
2018-01-26 10:05:26                          javascript-i/conditionals   
2018-01-26 10:05:36                             javascript-i/functions   
2018-01-26 10:08:22                           mkdocs/search_index.json   
2018-01-26 10:08:34                            git/merge-conflict-demo   
2018-01-26 10:08:45                           mkdocs/search_index.json   
2018-01-26 10:08:55                               git/working-in-teams   
2018-01-26 10:09:15          javascript-i/introduction/primitive-types   
2018-01-26 10:09:24                javascript-i/introduction/operators   
2018-01-26 10:10:16                             javascript-i/functions   
2018-01-26 10:14:47                                                  /   
2018-01-26 10:14:51                                       javascript-i   
2018-01-26 10:14:53                                                  /   
2018-01-26 10:15:02                             javascript-i/functions   
2018-01-26 10:15:23                                       javascript-i   
2018-01-26 10:15:26                  javascript-i/javascript-with-html   
2018-01-26 10:15:28  javascript-i/introduction/working-with-data-ty...   
2018-01-26 10:17:14                  javascript-i/javascript-with-html   
2018-01-26 10:17:16                             javascript-i/functions   
2018-01-26 10:18:22                                                  /   
2018-01-26 10:18:24                                             spring   
2018-01-26 10:33:18                                       javascript-i   
2018-01-26 10:33:22                  javascript-i/javascript-with-html   
2018-01-26 10:33:24                             javascript-i/functions   
2018-01-26 10:33:25                             javascript-i/functions   
2018-01-26 10:34:57                             javascript-i/functions   
2018-01-26 10:35:00                             javascript-i/functions   
2018-01-26 10:35:08  javascript-i/introduction/working-with-data-ty...   
2018-01-26 10:35:12                  javascript-i/javascript-with-html   
2018-01-26 10:35:14                             javascript-i/functions   

                    user_id cohort_id            ip  id        name  \
date                                                                  
2018-01-26 09:55:03       1       8.0  97.105.19.61   1      Arches   
2018

3. Are there students who, when active, hardly access the curriculum? If so, what information do you have about these students?

So here I want to find the 'active' dates for students(